pkg/edgecache/finalizer/controller.go

```diff
@@ -0,0 +1,216 @@
+
+	klog.Infof("Starting Hydra Finalizer controller")
+	defer klog.Infof("Shutting down Hydra Finalizer controller")
+
+	if !cache.WaitForNamedCacheSync("hydrafinalizer", ctx.Done(), c.pvcListerSynced) {
+		return
+	}
+
+	klog.V(2).Infof("Waiting on workers")
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}
+
+	<-ctx.Done()
+	klog.Infof("Return Run")
+}
+
+func (c *Controller) runWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+		klog.Info("processed next item")
+	}
+}
+
+// processNextWorkItem deals with one pvcKey off the queue.  It returns false when it's time to quit.
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+	var err error
+
+	pvc := obj.(*v1.PersistentVolumeClaim)
+	err = c.processPVC(pvc)
+	if err == nil {
+		c.queue.Forget(obj)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("VolumeEvent %v failed with : %w", obj, err))
+	c.queue.AddRateLimited(obj)
+
+	return true
+}
+
+func (c *Controller) processPVC(pvc *v1.PersistentVolumeClaim) error {
+	pvName := pvc.Spec.VolumeName
+	klog.Infof("Processing PVC %s:%s for PV %s", pvc.Namespace, pvc.Name, pvName)
+	// track the duration
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished processing (%v)", pvName, time.Since(startTime))
+	}()
+
+	// get the latest pv object
+	pv, err := c.client.CoreV1().PersistentVolumes().Get(context.TODO(), pvName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		klog.Infof("%s not found, ignoring", pvName)
+		return nil
+	}
+	if err != nil {
+		klog.Error(err)
+		return err
+	}
+	// ignore pvs not using edgecache or which do not have the finalizer
+	if !util.ContainsString(pv.GetFinalizers(), finalizerName, nil) {
+		return nil
+	}
+	// wait for the PVC to be deleted before cleaning up the volume
+	if pvc.GetDeletionTimestamp() == nil {
+		klog.Infof("PVC is not being deleted, ignoring")
+		return nil
+	}
+	// We always add the finalizer from node plugin so might as well use an annotation here!
+	err = c.ecManager.DeleteVolume(pv.ObjectMeta.Annotations[accountAnnotation], pv.ObjectMeta.Annotations[containerAnnotation])
+	if err != nil {
+		return err
+	}
+	klog.Infof("Edge Cache volume deleted, removing finalizer")
+	// edge cache delete success, remove the finalizer
+	return RemoveFinalizer(c.client, pv, pvc)
+}
+
+func (c *Controller) pvcUpdated(a interface{}, b interface{}) {
+	newpvc, ok := b.(*v1.PersistentVolumeClaim)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("PVC informer returned non-PVC object: %#v", b))
+		return
+	}
+	// ignore pvcs not using edgecache or which already have no finalizer
+	if !util.ContainsString(newpvc.GetFinalizers(), finalizerName, nil) {
+		return
+	}
+	// ignore any pvcs which still have attached pods
+	if util.ContainsString(newpvc.GetFinalizers(), pvcProtectionFinalizer, nil) {
+		return
+	}
+
+	/*
```

Comment:
is this dead code?  should it be removed?

Reply:
the json logging code you mean?

Reply:
lines 207 through 214

Reply:
removed